### PR TITLE
Add listFilesOnly command-line option

### PR DESCRIPTION
**Pull request by amcasey**
Msbuild currently uses a combination of `--listFiles` and `--noEmit` to
obtain a list of files to be consumed by tsc.  However, these two flags
don't suppress type checking, the results of which msbuild will never
consume.  This new switch gives msbuild a faster way to obtain the file
list.

Note: like `--noEmit`, doesn't make sense in build mode.

### CODE REVIEW

File: src/compiler/commandLineParser.ts [hunk 1]
```diff
@@ -216,6 +216,13 @@ namespace ts {
             isCommandLineOnly: true,
             description: Diagnostics.Print_the_final_configuration_instead_of_building
         },
+        {
+            name: "listFilesOnly",
+            type: "boolean",
+            category: Diagnostics.Command_line_Options,
```
Comment (amcasey) [inline src/compiler/commandLineParser.ts]: Does this "affect" semantic diagnostics?  Emit?

File: src/tsc/tsc.ts [hunk 2]
```diff
@@ -87,6 +87,11 @@ namespace ts {
             return sys.exit(ExitStatus.Success);
         }
 
+        if (commandLine.options.watch && commandLine.options.listFilesOnly) {
```
Comment (amcasey) [inline src/tsc/tsc.ts]: I didn't see any existing tests for this level of validation.

File: tests/baselines/reference/showConfig/Shows tsconfig for single option/listFilesOnly/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+    "compilerOptions": {
+        "listFilesOnly": true
```
Comment (amcasey) [inline tests/baselines/reference/showConfig/Shows tsconfig for single option/listFilesOnly/tsconfig.json]: I wouldn't have guessed such a file would be created for a command-line-only option.

File: tests/cases/fourslash/listFilesOnly.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+/// <reference path='fourslash.ts' />
+
+// @listFilesOnly: true
+
+// @Filename: a.js
+////const x = 1;
+
+// Just want to see that no baseline is emitted
+
+verify.noErrors();
```
Comment (amcasey) [inline tests/cases/fourslash/listFilesOnly.ts]: Fourslash throws if there isn't a call here.

File: src/compiler/program.ts [hunk 5]
```diff
@@ -1582,11 +1583,11 @@ namespace ts {
         function emitWorker(program: Program, sourceFile: SourceFile | undefined, writeFileCallback: WriteFileCallback | undefined, cancellationToken: CancellationToken | undefined, emitOnlyDtsFiles?: boolean, customTransformers?: CustomTransformers, forceDtsEmit?: boolean): EmitResult {
             let declarationDiagnostics: readonly Diagnostic[] = [];
 
-            if (!forceDtsEmit) {
-                if (options.noEmit) {
-                    return { diagnostics: declarationDiagnostics, sourceMaps: undefined, emittedFiles: undefined, emitSkipped: true };
-                }
+            if (options.listFilesOnly || (!forceDtsEmit && shouldSuppressEmit(options))) {
```
Comment (amcasey) [inline src/compiler/program.ts]: @sheetalkamat Is this okay?  This was the only place my new switch diverged from `--noEmit`.

File: src/compiler/program.ts [hunk 6]
```diff
@@ -1560,6 +1560,7 @@ namespace ts {
         }
 
         function getDiagnosticsProducingTypeChecker() {
+            Debug.assert(!options.listFilesOnly || !!options.extendedDiagnostics);
```
Comment (amcasey) [inline src/compiler/program.ts]: @RyanCavanaugh is there something softer than an assert that I could/should use here?  Asserting is great locally, but telemetry would suffice on a user's box.

File: src/compiler/commandLineParser.ts (hunk 1 above)
Comment (amcasey) [inline src/compiler/commandLineParser.ts]: @uniqueiniquity As far as I could tell, vstsc doesn't use this in build mode.  Does that sound right?

Comment (uniqueiniquity) [reply src/compiler/commandLineParser.ts]: Yes, it does. The idea is that in build mode, you should be depending on the incremental behavior of build mode itself, not our MSBuild version of it.

Comment (amcasey) [reply src/compiler/commandLineParser.ts]: To confirm, that's "Yes, it does sound right" and not "Yes, it does use the flag in build mode"?

Comment (uniqueiniquity) [reply src/compiler/commandLineParser.ts]: Yes, the former - it does sound right.

File: src/compiler/commandLineParser.ts [hunk 7]
```diff
@@ -216,6 +216,13 @@ namespace ts {
             isCommandLineOnly: true,
             description: Diagnostics.Print_the_final_configuration_instead_of_building
         },
+        {
+            name: "listFilesOnly",
+            type: "boolean",
+            category: Diagnostics.Command_line_Options,
+            isCommandLineOnly: true,
+            description: Diagnostics.Print_names_of_files_that_are_part_of_the_compilation_and_then_stop_processing
```
Comment (sheetalkamat) [inline src/compiler/commandLineParser.ts]: We should also make this Advanced options instead.

**Comment by amcasey** [comment]
@sheetalkamat That was my first attempt and it didn't work, but that might have been because I was trying to handle build mode.  I'll try again.

File: src/compiler/commandLineParser.ts (hunk 7 above)
Comment (amcasey) [reply src/compiler/commandLineParser.ts]: Ignorant question: what's the difference?

**Comment by amcasey** [comment]
@sheetalkamat I moved nearly all of the changes into `emitFilesAndReportErrors`, but I couldn't figure out why the options should be "advanced", rather than "command-line only".

File: src/compiler/commandLineParser.ts (hunk 1 above)
Comment (amcasey) [reply src/compiler/commandLineParser.ts]: Yes.

File: tests/baselines/reference/showConfig/Shows tsconfig for single option/listFilesOnly/tsconfig.json (hunk 3 above)
Comment (amcasey) [reply tests/baselines/reference/showConfig/Shows tsconfig for single option/listFilesOnly/tsconfig.json]: But it is - it's for illustration purposes only.

File: src/compiler/program.ts (hunk 5 above)
Comment (amcasey) [reply src/compiler/program.ts]: Moot.

File: src/compiler/commandLineParser.ts (hunk 7 above)
Comment (amcasey) [reply src/compiler/commandLineParser.ts]: I think we concluded that `Command_line_Options` is fine.

**Comment by amcasey** [comment]
Force pushed a rebase.  `tsc.js` changes are now in `executeCommandLine.ts`.

File: src/compiler/program.ts (hunk 6 above)
Comment (amcasey) [reply src/compiler/program.ts]: Removed from 3.7.  Will be revisited in 3.8 (possibly as a telemetry-only assert).

File: tests/cases/fourslash/listFilesOnly.ts (hunk 4 above)
Comment (amcasey) [reply tests/cases/fourslash/listFilesOnly.ts]: Moot.

File: src/tsc/tsc.ts (hunk 2 above)
Comment (amcasey) [reply src/tsc/tsc.ts]: `verifyTsc` will do it.

File: src/testRunner/unittests/tsc/listFilesOnly.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+            scenario: "listFilesOnly",
+            subScenario: "before watch",
+            fs: () => loadProjectFromFiles({
+                "/src/test.ts": utils.dedent`
+                        export const x = 1;`,
+            }),
+            commandLineArgs: ["/src/test.ts", "--listFilesOnly", "--watch"]
+        });
+
+        verifyTsc({
```
Comment (sheetalkamat) [inline src/testRunner/unittests/tsc/listFilesOnly.ts]: You don’t need two tests.. parsed command line handles that so one is enough

**Comment by amcasey** [comment]
Offline, @sheetalkamat said she was fine once the tests were fixed.

**Pull request merged**